### PR TITLE
implement framework of seq_pool_op and its unitest

### PR DESCRIPTION
**Pull request by luotao1**
solve part1 of #4186 
complete seq_sum_pool

### CODE REVIEW

File: paddle/operators/sequence_pool_op.cc [hunk 1]
```diff
@@ -42,21 +42,44 @@ class SequenceAvgPoolOp : public framework::OperatorWithKernel {
...
+        .SetDefault(AVERAGE)
+        .InEnum({AVERAGE, SUM, SQRT, MAX, LAST, FIRST});
     AddComment(R"DOC(
-    SequenceAvgPoolOp averages features of all time-steps of each instance.
-    More detailed comments will be added later.
+    SequencePoolOp pools features of all time-steps of each instance.
+
+    For a mini-batch of 3 variable lengths sentences, containing 2, 3, and 2 words:
+
+    X = [[1, 3], [2, 4, 6], [5, 1]],
```
Comment (qingqing01) [inline paddle/operators/sequence_pool_op.cc]: 换成：containing 2, 3, and 2 time-steps?

`X`换成`[7, M, N]`的float tensor？  例子更通用些

File: paddle/operators/sequence_pool_op.h [hunk 2]
```diff
@@ -28,54 +28,85 @@ template <typename T, int MajorType = Eigen::RowMajor,
...
+
+      switch (strategy) {
+        case AVERAGE:
+          out_e.device(place) = in_e.mean(Eigen::array<int, 1>({{0}}));
+          break;
+        case SUM:
+          out_e.device(place) = in_e.sum(Eigen::array<int, 1>({{0}}));
+          break;
+        default:
+          LOG(FATAL) << "unsupported pooling strategy";
```
Comment (qingqing01) [inline paddle/operators/sequence_pool_op.h]: 不要使用`LOG(FATAL)`， 用`PADDLE_THROW`:

```cpp
PADDLE_THROW("unsupported pooling strategy");
```

File: paddle/operators/sequence_pool_op.h [hunk 3]
```diff
@@ -28,54 +28,85 @@ template <typename T, int MajorType = Eigen::RowMajor,
...
+
+      switch (strategy) {
+        case AVERAGE:
+          in_g_e.device(place) = (out_g_e / static_cast<T>(h)).broadcast(bcast);
+          break;
+        case SUM:
+          in_g_e.device(place) = (out_g_e).broadcast(bcast);
+          break;
+        default:
+          LOG(FATAL) << "unsupported pooling strategy";
```
Comment (qingqing01) [inline paddle/operators/sequence_pool_op.h]: same as above.

File: paddle/operators/sequence_pool_op.cc (hunk 1 above)
Comment (luotao1) [reply paddle/operators/sequence_pool_op.cc]: Done

File: paddle/operators/sequence_pool_op.h (hunk 2 above)
Comment (luotao1) [reply paddle/operators/sequence_pool_op.h]: Done

File: paddle/operators/sequence_pool_op.h (hunk 3 above)
Comment (luotao1) [reply paddle/operators/sequence_pool_op.h]: Done

File: paddle/operators/sequence_pool_op.cc [hunk 4]
```diff
@@ -42,21 +42,45 @@ class SequenceAvgPoolOp : public framework::OperatorWithKernel {
...
  public:
-  SequenceAvgPoolOpMaker(framework::OpProto* proto,
-                         framework::OpAttrChecker* op_checker)
+  SequencePoolOpMaker(framework::OpProto* proto,
+                      framework::OpAttrChecker* op_checker)
       : OpProtoAndCheckerMaker(proto, op_checker) {
-    AddInput("X", "Input of SequenceAvgPoolOp.");
-    AddOutput("Out", "The output of SequenceAvgPoolOp.");
+    AddInput("X",
+             "A float LoDTensor, the variable-length input of SequencePoolOp");
```
Comment (qingqing01) [inline paddle/operators/sequence_pool_op.cc]: remove float

File: paddle/operators/sequence_pool_op.cc [hunk 5]
```diff
@@ -42,21 +42,45 @@ class SequenceAvgPoolOp : public framework::OperatorWithKernel {
...
+    AddOutput(
+        "Out",
+        "A float LoDTensor, the variable-length output of SequencePoolOp.");
```
Comment (qingqing01) [inline paddle/operators/sequence_pool_op.cc]: remove float

File: paddle/operators/sequence_pool_op.cc [hunk 6]
```diff
@@ -42,21 +42,45 @@ class SequenceAvgPoolOp : public framework::OperatorWithKernel {
...
+        "strategy",
+        "(int, default AVERAGE) the pooling strategy of SequencePoolOp.")
+        .SetDefault(AVERAGE)
+        .InEnum({AVERAGE, SUM, SQRT, MAX, LAST, FIRST});
     AddComment(R"DOC(
-    SequenceAvgPoolOp averages features of all time-steps of each instance.
-    More detailed comments will be added later.
+    SequencePoolOp pools features of all time-steps of each instance.
+
+    For a mini-batch of 3 variable lengths sentences, containing 2, 3, and 2 time-steps:
```
Comment (qingqing01) [inline paddle/operators/sequence_pool_op.cc]: variable-length sentences

File: paddle/operators/sequence_pool_op.cc [hunk 7]
```diff
@@ -42,21 +42,45 @@ class SequenceAvgPoolOp : public framework::OperatorWithKernel {
...
+    
+    Assume X is a [7,M,N] float LoDTensor, and X->lod()[0] = [0, 2, 5, 7].
```
Comment (qingqing01) [inline paddle/operators/sequence_pool_op.cc]: and X->lod()[0] = [0, 2, 5, 7],  7 = 2 +3 + 2

**Comment by luotao1** [comment]
Thx @qingqing01 . I will refine doc in next PR.

**Pull request merged**